File: MyApplication/library/src/main/java/com/a65apps/kalimullinilnazrafilovich/library/applicaiton/repositories/ContactLocationRepository.java
package com.a65apps.kalimullinilnazrafilovich.library.applicaiton.repositories;

import android.content.Context;
import android.util.Pair;

import androidx.annotation.NonNull;

import com.a65apps.kalimullinilnazrafilovich.entities.Contact;
import com.a65apps.kalimullinilnazrafilovich.entities.Location;
import com.a65apps.kalimullinilnazrafilovich.entities.Point;
import com.a65apps.kalimullinilnazrafilovich.interactors.details.ContactDetailsRepository;
import com.a65apps.kalimullinilnazrafilovich.interactors.location.LocationRepository;
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.db.AppDatabase;
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.mapper.YandexDTOMapper;
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.models.LocationOrm;
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.services.YandexGeocodeService;
import com.a65apps.kalimullinilnazrafilovich.myapplication.R;

import java.util.List;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public class ContactLocationRepository implements LocationRepository {
    private final AppDatabase database;
    private final Context context;
    private final ContactDetailsRepository contactDetailsContentResolverRepository;

    public ContactLocationRepository(@NonNull AppDatabase database,
                                     @NonNull Context context,
                                     @NonNull ContactDetailsRepository contactDetailsContentResolverRepository) {
        this.database = database;
        this.context = context;
        this.contactDetailsContentResolverRepository = contactDetailsContentResolverRepository;
    }

    @Override
    public void insertContactLocation(@NonNull Location location, @NonNull Contact contact) {
        LocationOrm locationORM = new LocationOrm(contact, location);
        database.locationDao().insert(locationORM);
    }

    @Override
    @NonNull
    public Single<List<Location>> getAllContactLocations() {
        return Single.fromCallable(() -> database.locationDao().getAll())
                .flatMapObservable(Observable::fromIterable)
                .flatMapSingle(item ->
                        contactDetailsContentResolverRepository.getDetailsContact(item.getContactID())
                                .map(contact -> new Pair<LocationOrm, Contact>(item, contact)))
                .map(pair -> new Location(
                        pair.second,
                        pair.first.getAddress(),
                        new Point(pair.first.getLatitude(), pair.first.getLongitude())
                )).toList();
    }

    @Override
    @NonNull
    public Single<Location> createOrUpdateContactLocationByCoordinate(@NonNull Contact contact,
                                                                      double latitude,
                                                                      double longitude) {
        String coordinate = latitude + "," + longitude;

        Single<Location> locationSingle = YandexGeocodeService.getInstance()
                .getJSONApi()
                .getLocation(
                        coordinate,
                        context.getResources().getString(R.string.yandex_maps_key))
                .map((responseDTO) -> new YandexDTOMapper().transform(
                        contact,
                        responseDTO,
                        latitude,
                        longitude))
                .doOnSuccess(location -> insertContactLocation(location, contact));

        return locationSingle;
    }
}

File: library/src/main/java/com/a65apps/kalimullinilnazrafilovich/library/applicaiton/repositories/ContactLocationRepository.kt
package com.a65apps.kalimullinilnazrafilovich.library.applicaiton.repositories

import android.content.Context
import com.a65apps.kalimullinilnazrafilovich.entities.ContactDetailsInfo
import com.a65apps.kalimullinilnazrafilovich.entities.Location
import com.a65apps.kalimullinilnazrafilovich.entities.Point
import com.a65apps.kalimullinilnazrafilovich.interactors.details.ContactDetailsRepository
import com.a65apps.kalimullinilnazrafilovich.interactors.location.LocationRepository
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.db.AppDatabase
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.mapper.YandexDTOMapper
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.models.LocationOrm
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.models.YandexAddressResponseDTO
import com.a65apps.kalimullinilnazrafilovich.library.applicaiton.services.YandexGeocodeService
import com.a65apps.kalimullinilnazrafilovich.myapplication.R
import io.reactivex.rxjava3.core.Single
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class ContactLocationRepository @Inject constructor(
    private val db: AppDatabase,
    private val context: Context,
    private val contactDetailsRepository: ContactDetailsRepository
) : LocationRepository {

    override fun insertContactLocation(location: Location, contactDetailsInfo: ContactDetailsInfo) {
        val locationOrm = LocationOrm(contactDetailsInfo, location)
        db.locationDao().insert(locationOrm)
    }

    override fun getAllContactLocations() = flow {
        emit(db.locationDao().getAll())
    }.map {
        toListLocation(it)
    }

    override fun createOrUpdateContactLocationByCoordinate(
        contactDetailsInfo: ContactDetailsInfo,
        latitude: Double,
        longitude: Double
    ): Single<Location> = YandexGeocodeService.getInstance()
        .jsonApi
        .getLocation(
            "$latitude,$longitude",
            context.resources.getString(R.string.yandex_maps_key)
        )
        .map { responseDTO: YandexAddressResponseDTO? ->
            YandexDTOMapper().transform(
                contactDetailsInfo,
                responseDTO!!,
                latitude,
                longitude
            )
        }
        .doOnSuccess { location: Location -> insertContactLocation(location, contactDetailsInfo) }

    private fun toListLocation(it: List<LocationOrm>) =
        mutableListOf<Location>().apply {
            it.map {
                contactDetailsRepository.getDetailsContact(it.contactID).map { contact ->
                    {
                        add(
                            Location(
                                contact,
                                it.address,
                                Point(
                                    it.latitude,
                                    it.latitude
                                )
                            )
                        )
                    }
                }
            }
        }
}
